core/rate-limit.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2014 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package core
+
+import (
+	"crypto"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"time"
+)
+
+type queue struct {
+	size  int
```

Comment:
Your queue needs a semaphore to protect all of its methods. It's probably best to do it on a per-queue basis, and make them fully synchronized (all methods).